### PR TITLE
Add price rule attributes when hydrating product during price calcula…

### PR DESCRIPTION
**Pull request by scazz010**
…tion

### Description
Column in `catalog_eav_attribute` is `is_used_for_price_rules`. This PR means that all attributes used for price calculation are fetched by the `lowestPriceOptionsProvider()` (which is called inside `module-configurable-product/Pricing/Price/ConfigurablePriceResolver.php`->`resolvePrice`) 

Backstory / justification: 

I was adding a feature to control price visibility (for some products, we don't want to display the price on our magento store). This includes for some options of configurable products.  Despite setting `is_used_for_price_rules` to be true, the product model is hydrated with only some default attributes (special prices, tax class), but we need to be able to add some extra processing to check we're allowed to display to price

### Manual testing scenarios
1. Create a new attribute, for example, `price_visibility`, and ensure that it's corresponding entry in `catalog_eav_attribute` has `is_used_for_price_rules` to be true
2. Add this attribute to a configurable product's attribute set, and set some of the simple products to have this as false
3. Create a new configurable price resolver and attempt to filter sub products so that only those with price_visibility of true are considered. 

### Contribution checklist
 - [x] Pull request has a meaningful description of its purpose
 - [x] All commits are accompanied by meaningful commit messages
 - [x] All new or changed code is covered with unit/integration tests (if applicable)
 - [x] All automated tests passed successfully (all builds on Travis CI are green)

### CODE REVIEW

**Comment by magento-cicd2** [comment]
[CLA assistant check](https://cla.dev.magento.com/magento/magento2?pullRequest=12423) <br/>All committers have signed the CLA.

**Comment by avoelkl** [comment]
Hi @scazz010 
Thanks for your PR! There is a failing codestyle check - could you please have a look and fix it? Thanks!

**Comment by scazz010** [comment]
@avoelkl Sure, have fixed!

File: app/code/Magento/Catalog/Model/Config.php [hunk 1]
```diff
@@ -64,6 +64,13 @@ class Config extends \Magento\Eav\Model\Config
      */
     protected $_usedForSortBy;
 
+    /**
+     * Product Attributes For Price Rules
+     *
+     * @var array
+     */
+    protected $_usedForPriceRules;
```
Comment (ihor-sviziev) [inline app/code/Magento/Catalog/Model/Config.php:72]: 1. Please change property visibility from protected to private
2. Please remove underscore for variable name

File: app/code/Magento/ConfigurableProduct/Pricing/Price/LowestPriceOptionsProvider.php [hunk 2]
```diff
@@ -37,19 +38,37 @@ class LowestPriceOptionsProvider implements LowestPriceOptionsProviderInterface
...
      * @param LinkedProductSelectBuilderInterface $linkedProductSelectBuilder
      * @param CollectionFactory $collectionFactory
+     * @param Config $catalogConfig
      */
     public function __construct(
         ResourceConnection $resourceConnection,
         LinkedProductSelectBuilderInterface $linkedProductSelectBuilder,
-        CollectionFactory $collectionFactory
+        CollectionFactory $collectionFactory,
+        Config $catalogConfig
```
Comment (ihor-sviziev) [inline app/code/Magento/ConfigurableProduct/Pricing/Price/LowestPriceOptionsProvider.php:56]: This change is not backward compatible. Please make this parameter optional with fallback of getting this object instance from objectManager

File: app/code/Magento/ConfigurableProduct/Pricing/Price/LowestPriceOptionsProvider.php [hunk 3]
```diff
@@ -37,19 +38,37 @@ class LowestPriceOptionsProvider implements LowestPriceOptionsProviderInterface
...
         $this->collectionFactory = $collectionFactory;
+        $this->catalogConfig = $catalogConfig;
+    }
+
+    /**
+     * Load Product attributes
+     *
+     * @return array
+     */
+    public function getPriceRuleAttributes()
```
Comment (ihor-sviziev) [inline app/code/Magento/ConfigurableProduct/Pricing/Price/LowestPriceOptionsProvider.php:69]: Could you change visibility to private?

**Review by ihor-sviziev — CHANGES_REQUESTED** [review]

**Comment by ihor-sviziev** [comment]
Hi @scazz010,
Are you interested in finalizing this PR?

**Comment by sidolov** [comment]
@scazz010 , I am closing this PR now due to inactivity.
Please reopen and update if you wish to continue.
Thank you for collaboration!

**Pull request closed without merging**